typescript-requiring-type-checking.js

```diff
@@ -15,23 +15,23 @@ module.exports = {
         /*
```

Comment:
Discussion for recommended rule: `'@typescript-eslint/no-floating-promises': 'error'`

Reply:
I ran into several errors with [@typescript-eslint/no-floating-promises](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-floating-promises.md) that are worth discussing.

1. In Jasmine tests, assert functions return `Promise<void>` so lines like this violate the rule: `expect(serialized.widgets).toBeDefined();`.
   - possible mitigation: turn this off for test code
   - ugly mitigation from [StackOverflow](https://stackoverflow.com/questions/56938854/how-to-ignore-jasmine-promises-when-no-floating-promises-in-tslint-is-true): add the `void` keyword before every assert. `void expect(serialized.widgets).toBeDefined();`
1. Angular's `router.navigate()` function returns a `Promise<boolean>` that describes whether the navigation was successful but [none of NI's apps](https://dev.azure.com/ni/DevCentral/_search?action=contents&text=router.navigate%20ext%3Ats&type=code&lp=code-Project&filters=ProjectFilters%7BDevCentral%7D&pageSize=25&result=DefaultCollection%2FDevCentral%2Fsrd-semi-ses-software%2FGBmain%2F%2FSpecManagement%2Fsrc%2Fapp%2Fauthentication%2Fauth.guard.ts) await or otherwise handle that promise.
1. In [production code](https://dev.azure.com/ni/DevCentral/_git/op-web-portal?path=%2FOptimalTest.WebPortal%2Fsrc%2Fapp%2Fapp.component.ts&version=GBmain&line=42&lineEnd=43&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents), calling an async function without awaiting it. This seems like a legit error and we should fix it by awaiting or explicitly saying we don't need it to complete (via `void`?)
1. [Some code](https://dev.azure.com/ni/DevCentral/_git/op-web-portal?path=%2FOptimalTest.WebPortal%2Fsrc%2Fapp%2Fdashboard-catalog%2Fcomponents%2Fdashboard-catalog%2Fdashboard-catalog.component.ts&version=GBmain&line=267&lineEnd=267&lineStartColumn=29&lineEndColumn=51&lineStyle=plain&_a=contents) handles a promise via `then()` but only handles the `resolve` case but not the `reject`. Would our recommendation be to always provide a `reject` callback?

Reply:
Point of clarification on 1. since Milan was asking about it offline. The functions that return a promise are the "matchers" (`toBeDefined()`, `toEqual()`, etc). This [behavior](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/0b5bfba2994c91a099cd5bcfd984f6c4c39228e5/types/jasminewd2/index.d.ts) is specific to [jasminewd2](https://www.npmjs.com/package/jasminewd2), which is a version of Jasmine used by Protractor (and used by default in Angular apps) with async enhancements for Web Driver.

Reply:
General consensus at our meeting today was that the rule is providing some value in cases 2, 3, and 4. Case 1 is unfortunate and we might provide guidance that the rule can be disabled for test code that uses `jasminewd2`. I'll leave this thread open for now but since it was moderately controversial it should be resolved by voting on a proposal.

Reply:
Proposal: Leave the rule enabled and add a comment similar to "Angular projects with tests leveraging the `jasminewd2` library may choose to disable this rule for test code but should consider keeping it enabled for source."

Reply:
We're all in agreement that this is a good rule. 
@rajsite will double check whether this applies only to protractor tests or also to Jasmine tests. If the former, no special comment is needed. If the latter, we should keep the comment and maybe add an example about how to suppress the rule for certain files.

Reply:
Looks like it effects both test sweets. This is the comment I added:

```js
        // Notes for recommended configuration: '@typescript-eslint/no-floating-promises': 'error'
        // The jasminewd2 library used by Angular applications will result in a significant number of floating promises.
        // Consider disabling the '@typescript-eslint/no-floating-promises' rule for tests that utilize jasmine expects.
```